addons/l10n_eg_edi_eta/models/account_edi_format.py

```diff
@@ -0,0 +1,379 @@
+                  "l10n_eg_building_no"]
+        if (invoice and invoice.amount_total >= invoice.company_id.l10n_eg_invoicing_threshold) or self._l10n_eg_get_partner_tax_type(partner_id, issuer) != 'P':
+            fields.append('vat')
+        return all(partner_id[field] for field in fields)
+
+    @api.model
+    def _l10n_eg_eta_prepare_eta_invoice(self, invoice):
+
+        def group_tax_retention(tax_values):
+            return {'l10n_eg_eta_code': tax_values['tax_id'].l10n_eg_eta_code.split('_')[0]}
+
+        date_string = invoice.invoice_date.strftime('%Y-%m-%dT%H:%M:%SZ')
+        grouped_taxes = invoice._prepare_edi_tax_details(grouping_key_generator=group_tax_retention)
+        invoice_line_data, totals = self._l10n_eg_eta_prepare_invoice_lines_data(invoice, grouped_taxes['invoice_line_tax_details'])
+        eta_invoice = {
+            'issuer': self._l10n_eg_eta_prepare_address_data(invoice.journal_id.l10n_eg_branch_id, invoice, issuer=True,),
+            'receiver': self._l10n_eg_eta_prepare_address_data(invoice.partner_id, invoice),
+            'documentType': 'i' if invoice.move_type == 'out_invoice' else 'c' if invoice.move_type == 'out_refund' else 'd' if invoice.move_type == 'in_refund' else '',
+            'documentTypeVersion': '1.0',
+            'dateTimeIssued': date_string,
+            'taxpayerActivityCode': invoice.journal_id.l10n_eg_activity_type_id.code,
+            'internalID': invoice.name,
+        }
+        eta_invoice.update({
+            'invoiceLines': invoice_line_data,
+            'taxTotals': [{
+                'taxType': tax['l10n_eg_eta_code'].split('_')[0].upper(),
+                'amount': self._l10n_eg_edi_round(abs(tax['tax_amount'])),
+            } for tax in grouped_taxes['tax_details'].values()],
+            'totalDiscountAmount': self._l10n_eg_edi_round(totals['discount_total']),
+            'totalSalesAmount': self._l10n_eg_edi_round(totals['total_price_subtotal_before_discount']),
+            'netAmount': self._l10n_eg_edi_round(abs(invoice.amount_untaxed_signed)),
+            'totalAmount': self._l10n_eg_edi_round(abs(invoice.amount_total_signed)),
+            'extraDiscountAmount': 0.0,
+            'totalItemsDiscountAmount': 0.0,
+        })
+        return eta_invoice
+
+    @api.model
+    def _l10n_eg_eta_prepare_invoice_lines_data(self, invoice, tax_data):
+        lines = []
+        totals = {
+            'discount_total': 0.0,
+            'total_price_subtotal_before_discount' : 0.0,
+        }
+        for line in invoice.invoice_line_ids.filtered(lambda x: not x.display_type):
+            line_tax_details = tax_data.get(line, {})
+            price_unit = self._l10n_eg_edi_round(abs((line.balance / line.quantity) / (1 - (line.discount / 100.0)))) if line.quantity and line.discount != 100.0 else line.price_unit
+            price_subtotal_before_discount = self._l10n_eg_edi_round(abs(line.balance / (1 - (line.discount / 100))))
```

Comment:
then here we should have if line.discount as well else line.quantity * line.price_unit?  @akr-odoo